components/logins/src/error.rs

```diff
@@ -97,193 +112,52 @@ pub enum InvalidLogin {
-                LoginsStorageError::InvalidKey(label)
-            }
-
-            ErrorKind::InvalidKey => LoginsStorageError::InvalidKey(label),
-
-            ErrorKind::SqlError(rusqlite::Error::SqliteFailure(err, _))
-                if err.code == rusqlite::ErrorCode::OperationInterrupted =>
-            {
-                log::warn!("Operation interrupted (SQL)");
-                LoginsStorageError::Interrupted(label)
+            // Our internal "no such record" error is converted to our public "no such record" error, with no logging and no error reporting.
+            Self::NoSuchRecord(guid) => {
+                Self::passthrough(LoginsStorageError::NoSuchRecord(guid.to_string()))
             }
-
-            ErrorKind::Interrupted(_) => {
-                log::warn!("Operation interrupted (Outside SQL)");
-                LoginsStorageError::Interrupted(label)
-            }
-
-            ErrorKind::InvalidSalt => {
-                log::error!("Invalid salt provided");
-                // In the old world, this had an error code (7) but no Kotlin
-                // error type, meaning it got the "base" error.
-                LoginsStorageError::UnexpectedLoginsStorageError(label)
-            }
-
-            ErrorKind::CryptoError(_) => {
-                log::error!("Crypto error");
-                LoginsStorageError::CryptoError(label)
-            }
-
-            err => {
-                log::error!("UnexpectedLoginsStorageError error: {:?}", err);
-                LoginsStorageError::UnexpectedLoginsStorageError(label)
+            // NonEmptyTable error is just a sanity check to ensure we aren't asked to migrate into an
+            // existing DB - consumers should never actually do this, and will never expect to handle this as a specific
+            // error - so it gets reported to the error reporter and converted to an "internal" error.
+            Self::NonEmptyTable => Self::unexpected(
+                LoginsStorageError::UnexpectedLoginsStorageError(
+                    "must be an empty DB to migrate".to_string(),
+                ),
+                Some("migration"),
+            ),
+            Self::CryptoError(_) => Self::log(LoginsStorageError::IncorrectKey, log::Level::Warn),
+            Self::Interrupted(_) => {
+                Self::passthrough(LoginsStorageError::Interrupted(self.to_string()))
             }
+            // This list is partial - not clear if a best-practice should be to ask that every
+            // internal error is listed here (and remove this default branch) to ensure every error
+            // is considered, or whether this default is fine for obscure errors?
+            // But it's fine for now because errors were always converted with a default
+            // branch to "unexpected"
+            _ => Self::unexpected(
+                LoginsStorageError::UnexpectedLoginsStorageError(self.to_string()),
+                None,
+            ),
```

Comment:
I like the mechanics of how this works, but it seems like it could be simplified some.  What do you think of these suggestions:

 - Splitting up this method into `get_error_reporting()` and `into_external_error()`.  I think that would clarify what's getting reported and what's getting passed on to the consumer app.
  - Defining `report()`, `unexpected()`, etc. as methods of `ErrorHandling` (or `ErrorReporting` if we make the previous change).  This feels more natural to me, since those methods are just creating instances of those types.

Reply:
> Splitting up this method into get_error_reporting() and into_external_error(). I think that would clarify what's getting reported and what's getting passed on to the consumer app.

My concern with this is that it I fear it might be error prone - logins would then have 2 different functions which match every "internal" error - so it seems like it would be easier to accidentally have some errors (say) returning an InternalError, but not logging it, because you made an assumption about what the "other" match is going some 50 lines away. The way things stand, now there's a single place to get the whole context for a single error.

IOW, I thought the single match was quite desirable. I don't feel too strongly about this though, so welcome other thoughts. 

> Defining report(), unexpected(), etc. as methods of ErrorHandling

Yeah, this does make more sense, thanks.

Reply:
> My concern with this is that it I fear it might be error prone - logins would then have 2 different functions which match every "internal" error - so it seems like it would be easier to accidentally have some errors (say) returning an InternalError, but not logging it, because you made an assumption about what the "other" match is going some 50 lines away. The way things stand, now there's a single place to get the whole context for a single error.

That makes sense to me.  Maybe there's a way to keep everything in 1 match and also clarify what's happening.  Now that the `ErrorReporting` struct is using options, [what about a builder-style API to create `ErrorHandler` instances?](https://github.com/mozilla/application-services/commit/96767f5d543e87687bae55f9a09191e6b65cdf24)